Handle the program entrypoint not having a return address

tetsuo-cpp opened the pull request:
Closes #250

@pgoodman 
This still needs testing. I'm not able to use the Anvill Python package even with my virtual env activated so I'm not sure what's going on there. Will figure this out tomorrow.

What I still need to do:
- [x] Test Anvill (without this patch) on a function without any returns. We don't expect it to fall over but we should verify.
- [x] Update IDA script to populate `has_return_address`
- [x] Figure out how to populate `has_return_address` when translating from an LLVM function.
- [x] Test with this change see whether the original bug where `anvill_ra` references were left in the IR is fixed.

pgoodman commented:
>  I'm not able to use the Anvill Python package even with my virtual env activated so I'm not sure what's going on there. 

Try this from within your virtual environment:
```shell
python /Applications/Binary\ Ninja.app/Contents/Resources/scripts/install_api.py -v
```

Also, possibly you need to configure with `-DANVILL_INSTALL_PYTHON3_LIBS=ON`.

>  Figure out how to populate has_return_address when translating from an LLVM function.

I think you can assume all LLVM functions have a return address. Yes this is also a thing where round-tripping would break a bit.

tetsuo-cpp commented:
I've done a bit of testing today. Anvill works ok for the Godbolt example you posted on #250 where we have a function with no return (I don't see any stray `anvill_ra`).

Since I've changed the specification, I can't use the JSON spec attached on the original issue so I've been playing around trying to see what triggers the original bug. I noticed that this example causes stray `anvill_ra`. I actually see it in the `strdup` body itself which is making me think my assumption is wrong. If it only happens because of a lack of return address, I'd expect that the way to repro would be to have code in `_start` which accesses something on the stack which Anvill incorrectly assumes to be the return address. But the fact I'm able to repro with this snippet suggests that it's something else.

```c
#include <string.h>
int x = 0;
int y = 1;

__attribute__((noinline))
int bar(void) {
    return x;
}

__attribute__((noinline))
int foo(void) {
    return y;
}

int main(int argc, char** argv) {
    char* arg_copy = strdup(argv[0]);
    return argc ? foo() : bar();
}
```

I'll have to continue looking tomorrow. In the meantime, should I also get a license for IDA? I'm able to test the Binja script and it works fine but I won't be able to verify any changes to the IDA script unless there's some shared license I can use.

pgoodman commented:
The "stray" `__anvill_ra` in `_start` is because `_start` takes its "arguments" on the stack. That is, the Linux kernel pushes a few things onto the stack, and then `_start` loads/pops them off. In a normal function, it would appear as though `_start` were reading its own return address; however, a return address is an artifact of a `call` instruction, and `_start` is not actually called by anything. So, what this "stray" is is that the return address is taken off the stack, but then passed as an argument to `__libc_start_main` or something like that.

tetsuo-cpp commented:
Trying to clear my thoughts... I think there are two separate issues I'm seeing that lead to `anvill_ra` being left in the lifted program.

The first is the example that prompted this issue where we see `anvill_ra` referenced in `_start` with no call to `__remill_function_return` since we don't return out of the entry point. As you said, it's caused by `_start` reading things off the stack (what would normally be a return address). The patch in this PR should fix this. I haven't verified whether it works because I haven't been able to write a program and generate a spec which triggers this locally (and since I've changed the spec format in this patch, I can't use the one linked in the issue). I'm assuming that it is specific to how Linux hands off arguments to the program entry point so I'm thinking I might have to setup a VM to see it.

The second is triggered by that `strdup` example I listed above and doesn't seem to have anything to do with `_start`. I see `anvill_ra` in both the `strdup` body and another function. There are a lot of instructions in each but there is a common pattern at the end that looks like this:
```llvm
  %1241 = load i64, i64* @__anvill_reg_TPIDR_EL0, align 8
  store i64 %1241, i64* %128, align 8
  %1242 = load i64, i64* @__anvill_reg_TPIDRRO_EL0, align 8
  store i64 %1242, i64* %130, align 8
  store i64 %18, i64* %119, align 16
  store i64 ptrtoint (i8* @__anvill_ra to i64), i64* %117, align 16
  %1243 = load i64, i64* bitcast (i64 ()** @var_100008000__Sv to i64*), align 8
  store i64 %1243, i64* %89, align 16, !tbaa !32
  store i64 %1243, i64* %121, align 16
  %1244 = call %struct.Memory* @__remill_jump(%struct.State* %19, i64 %1243, %struct.Memory* null)
  %1245 = load i64, i64* %57, align 16
  ret i64 %1245
```
So there's no `__remill_function_return` meaning that there wouldn't have been a `ret` in the original function and instead there is a `__remill_jump`. I noticed there's a pass specifically to convert `__remill_jump => __remill_function_return` so maybe that should be triggering. I'm doing this all on an ARM machine if that's relevant, I see a comment in that pass that says:

> The transform is written to fix the bitcode generated for aarch32 architecture

I've attached the source, binary, spec and text LLVM IR in a tarball if you're interested: [strdup_test.tar.gz](https://github.com/lifting-bits/anvill/files/7091151/strdup_test.tar.gz)

pgoodman commented:
I can reproduce the issue and now I understand it :-) So, here is `strdup`:

![image](https://user-images.githubusercontent.com/43664/131755301-5c0c858e-e47d-4fd1-a4ec-2fbf45779689.png)

It is a [thunk](https://github.com/lifting-bits/anvill/issues/5) used to call functions defined outside of the library/binary. What we have happening is the following:

The implementation of the thunk uses an indirect jump. We are lacking any kind of information about targets for that indirect jump, and so we lift it using a call to `__remill_jump`. But, this call takes a `State *` argument, and so we say that the state structure "escapes" this function. The state structure represents the link register `lr` where the return address is stored. The combination of these two facts means the following:

 - On entry to the function, we store `__anvill_ra` into the emulated `LR` in the `State` structure. This is guided by the spec, which tells us that return addresses on entry to functions in aarch64 are stored in `LR`.
 - The function does an indirect tail call (`br`) to the address of `strdup`, stored in a sort of variable named `_strdup`.
 - The indirect tail call uses a `__remill_jump`, which cases the `State` structure to escape.
 - Because the state structure escapes, so does the value stored into link register (`__anvill_ra`).

So, the "bug" here is that we really want to be identifying the `br` as having a single "known" target. There are some features in the spec format (that unfortunately don't seem to be documented) for representing this. There are redirections, e.g. so that we can say a call to `_strdup` is really a call to the external `strdup`, and there are target lists, where we can say "the targets of this `br` instruction are known to be X, Y, and Z."

tetsuo-cpp commented:
@pgoodman 
Cool! I think is almost good to go.

I've tested with the binary you provided with and without my patch and can confirm that the fix seems to be working. I had a look at the stack frame recovery code and it seems to explicitly check whether we have uses of `anvill_ra` so I don't think there are any problems there.

The only thing left is to add support to the IDA script to figure out whether a function is the program entrypoint. I don't currently have an IDA license so I'll have to figure that out.

tetsuo-cpp commented on anvill/python/anvill/binja/bnfunction.py:
```diff
@@ -38,10 +38,12 @@ def __init__(
         param_list,
         ret_list,
         func_type,
+        has_return_address,
```
f2f29d6

tetsuo-cpp commented on anvill/include/anvill/Decl.h:
```diff
@@ -173,6 +173,9 @@ struct FunctionDecl {
   // Is this a variadic function?
   bool is_variadic{false};
 
+  // Does this function have a return address?
```
a7bf35e

tetsuo-cpp commented:
I think this is ready. @pgoodman, could you please help me with the IDA Pro script?

pgoodman commented:
> I think this is ready. @pgoodman, could you please help me with the IDA Pro script?

@kumarak can you look into this when you have time?

kumarak commented:
Binja and IDA script has function [is_noreturn](https://github.com/lifting-bits/anvill/blob/master/anvill/python/anvill/binja/bnfunction.py#L66) that set the field `is_noreturn` in the spec if the function does not return. I think the field can be used while lifting.  It seems there is a bug and binja script fail to identify `_start` as noreturn function.

tetsuo-cpp commented:
> Binja and IDA script has function [is_noreturn](https://github.com/lifting-bits/anvill/blob/master/anvill/python/anvill/binja/bnfunction.py#L66) that set the field `is_noreturn` in the spec, if the function does not return. I think the field can be used.

@pgoodman has a comment about that field [here](https://github.com/lifting-bits/anvill/issues/250#issuecomment-907867783).

kumarak commented:
Thanks @tetsuo-cpp for the reference. I find name `has_return_address` confusing. Probably we can have return address type void or empty `return_address` if the function does not have return address (or we want to ignore it).   

Also, I see you only look for entry function to set `has_return_address` flag. Could there be other cases which we would like as function with no return address? Probably a call to `longjmp`? If this is only for the `entry_function` then I think we should just handle `_start` function differently as entry point.

pgoodman commented:
> Thanks @tetsuo-cpp for the reference. I find name `has_return_address` confusing. Probably we can have return address type void or empty `return_address` if the function does not have return address (or we want to ignore it).

@kumarak lets try doing the way you mentioned here, with the type.

> Also, I see you only look for entry function to set `has_return_address` flag. Could there be other cases which we would like to treat as function with no return address? Probably a call to `longjmp`?

I hope in practice we don't find ourselves lifting `longjmp` implementations. Let's ignore this for now.

> If this is only for the `entry_function` then I think we should just handle `_start` function differently as entry point.

I think this special handling should be done in the producers of specifications. We can just have the lifter not have to care what an entry function is or isn't.

tetsuo-cpp commented:
@pgoodman @kumarak 
Ok, I've addressed your comments in f45f88e142a469c1bdf41efd9ac3d11f38a83e44.

With this commit, a function specification will omit `return_address` if it is the program entrypoint. When we get to `anvill-decompile-json`, if I notice that a spec has no `return_address`, I set the type of the return address to be void and a check like this to figure out whether to initialise the return address.
```
if (!decl.return_address.type->isVoidTy())
```

kumarak reviewed (approved):
I will update the IDA script to handle function with no `return address`. The changes looks good.

kumarak commented:
@pgoodman @tetsuo-cpp, I updated IDA script to treat `_start` as function with no return address. If it looks good then I think the PR is ready to be merge.

tetsuo-cpp commented:
Thanks so much @kumarak! Those changes look good to me.

@pgoodman, if you agree, could you please merge this?
